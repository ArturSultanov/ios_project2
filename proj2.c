//Artur Sultanov
//xsulta01

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/sem.h>
#include <sys/shm.h>
#include <errno.h>
#include <fcntl.h>
#include <semaphore.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/mman.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <sys/types.h>


#define SEMAPHORE_MUTEX "/xsulta01_sem_mutex"
#define SEMAPHORE_SERVICEFRST "/xsulta01_servicefrst" 
#define SEMAPHORE_SERVICESCND "/xsulta01_servicescnd" 
#define SEMAPHORE_SERVICETHRD "/xsulta01_servicethrd" 
#define SEMAPHORE_CLERK "/xsulta01_clerk"
#define SEMAPHORE_CUSTOMER "/xsulta01_customer"

typedef struct {
    int NZ; //number of clients
    int NU; //number of workers
    int TZ; //Time in ms where client waiting to go for post
    int TU; //worker rest, in ms
    int F; //in ms, time after that post office will be closed  
}args_t;

#define NUM_SERVICES 3
#define upsleep_for_random_time(time_max) usleep((rand() % (time_max + 1)) * 1000)


// Function prototypes
void semaphore_dest(void);
int semaphore_init(void);
void shared_memory_dest(void);
int shared_memory_init(void);
void customer_process(int idZ, int TZ);
void clerk_process(int idU, int TU);
int random_number(int min, int max);

// Global values
FILE *file;

// Shared memory declaration
bool *post_is_closed;
int *first_service_queue;
int *second_service_queue;
int *third_service_queue;
int *action_number;
int *clerks_number;
int *customers_numbers;


// deklaracia semaforov
sem_t *sem_mutex;
sem_t *sem_first_service;
sem_t *sem_second_service;
sem_t *sem_third_service;
sem_t *sem_clerk;
sem_t *sem_customer;

////////////////////////////    MAIN START  ////////////////////////////
int main(int argc, char *argv[]) {
    // if (argc != 6) {
    //     fprintf(stderr, "Error: Invalid number of arguments.\n");
    //     return 1;
    // }

    // int NZ = atoi(argv[1]); //počet zákazníků
    // int NU = atoi(argv[2]); //počet úředníků
    // int TZ = atoi(argv[3]); 
    // int TU = atoi(argv[4]);
    // int F = atoi(argv[5]);


    int NZ = 30; //počet zákazníků
    int NU = 20; //počet úředníků
    int TZ = 20; 
    int TU = 20;
    int F = 20;

    // Check if input values are within allowed range
    if (NZ < 0 || NU < 0 || TZ < 0 || TZ > 10000 || TU < 0 || TU > 100 || F < 0 || F > 10000) {
        fprintf(stderr, "Error: Invalid input values.\n");
        return 1;
    }
    

    //pid_t wpid;
    //int status = 0;



    // Initialize shared memory and semaphores
    if(shared_memory_init()){
        fprintf(stderr, "Cannot alocate shared memory!\n");
        return 1;
    }

    if(semaphore_init()){
        fprintf(stderr, "Cannot open semaphores!\n");
        return 1;
    }

    // Set output file and output buffer
    if ((file = fopen("proj2.out", "w+")) == NULL) {
        fprintf(stderr, "ERROR: Output file could not be opened\n");
        return 1;
    }
    setbuf(file, NULL);
    setbuf(stdout, NULL);
    //setbuf(stderr, NULL);

    // Fork customer processes
    for(int idZ  = 1; idZ <=NZ;idZ++){
        pid_t pid = fork();
        if(pid==0){
            //printf("CUSTOMER\n");
            srand((int)time(0) % getpid()); // for upsleep_for_random_time(time_max)
            customer_process(idZ, TZ);
            exit(0);
        }
        else if (pid==-1){
            fprintf(stderr, "Fork customer processes error!\n");
            shared_memory_dest();
            semaphore_dest();
            fclose(file);
            exit(1);
        }
    }

    // Fork clerk processes
    for(int idU  = 1; idU <=NU;idU++){
        pid_t pid = fork();
        if(pid==0){
            //printf("CLERK\n");
            srand((int)time(0) % getpid()); // for upsleep_for_random_time(time_max)
            clerk_process(idU, TU);
            exit(0);
        }
        else if (pid==-1){
            fprintf(stderr, "Fork clerk processes error!\n");
            shared_memory_dest();
            semaphore_dest();
            fclose(file);
            exit(1);
        }
    }

    usleep((rand() % ((F / 2) + 1)) * 1000 + F / 2 * 1000);

   //Close the post office
    sem_wait(sem_mutex);
    (*post_is_closed) = true;
    fprintf(file, "%d: closing\n", ++(*action_number));
    sem_post(sem_mutex);

    // Clean up shared memory and semaphores
    //while ((wpid = wait(&status)) > 0);
    while(wait(NULL)>0);
    shared_memory_dest();
    semaphore_dest();
    fclose(file);

    return 0;
}
////////////////////////////    MAIN END    ////////////////////////////

// Semaphores destruction
void semaphore_dest(void){
    sem_close(sem_mutex);   
    sem_unlink(SEMAPHORE_MUTEX);

    sem_close(sem_first_service);   
    sem_unlink(SEMAPHORE_SERVICEFRST);

    sem_close(sem_second_service);   
    sem_unlink(SEMAPHORE_SERVICESCND);

    sem_close(sem_third_service);   
    sem_unlink(SEMAPHORE_SERVICEFRST);

    sem_close(sem_clerk);   
    sem_unlink(SEMAPHORE_CLERK);


    sem_close(sem_customer);   
    sem_unlink(SEMAPHORE_CUSTOMER);

    if (file != NULL) fclose(file);

    return;
}

// Semaphores initialization
int semaphore_init(void){
    semaphore_dest();

    sem_mutex = sem_open(SEMAPHORE_MUTEX, O_CREAT | O_EXCL, 0666, 1) ;
    if (sem_mutex == SEM_FAILED){
        return 1;
    }
    sem_first_service = sem_open(SEMAPHORE_SERVICEFRST, O_CREAT | O_EXCL, 0666, 0) ;
    if (sem_first_service == SEM_FAILED){
        return 1;
    }

    sem_second_service = sem_open(SEMAPHORE_SERVICESCND, O_CREAT | O_EXCL, 0666, 0) ;
    if (sem_second_service == SEM_FAILED){
        return 1;
    }

    sem_third_service = sem_open(SEMAPHORE_SERVICETHRD, O_CREAT | O_EXCL, 0666, 0) ;
    if (sem_third_service == SEM_FAILED){
        return 1;
    }

    sem_clerk = sem_open(SEMAPHORE_CLERK, O_CREAT | O_EXCL, 0666, 1) ;
    if (sem_clerk == SEM_FAILED){
        return 1;
    }

    sem_customer = sem_open(SEMAPHORE_CUSTOMER, O_CREAT | O_EXCL, 0666, 0) ;
    if (sem_customer == SEM_FAILED){
        return 1;
    }

    return 0;
}

// Shared memory destruction
void shared_memory_dest(void){

    munmap(post_is_closed, sizeof(bool));
    munmap(first_service_queue, sizeof(int));
    munmap(second_service_queue, sizeof(int));
    munmap(third_service_queue, sizeof(int));
    munmap(action_number, sizeof(int));
    munmap(clerks_number, sizeof(int));
    munmap(customers_numbers, sizeof(int));

    return;
}

// Shared memory initialization
int shared_memory_init(void){
    void shared_memory_dest(void);

    post_is_closed = mmap(NULL, sizeof(bool), PROT_READ | PROT_WRITE,  MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if ( MAP_FAILED == post_is_closed) {
        return 1;
    }

    first_service_queue = mmap(NULL, sizeof(bool), PROT_READ | PROT_WRITE,  MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if ( MAP_FAILED == first_service_queue) {
        return 1;
    }
    second_service_queue = mmap(NULL, sizeof(bool), PROT_READ | PROT_WRITE,  MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if ( MAP_FAILED == second_service_queue) {
        return 1;
    }
    third_service_queue = mmap(NULL, sizeof(bool), PROT_READ | PROT_WRITE,  MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if ( MAP_FAILED == third_service_queue) {
        return 1;
    }

    action_number = mmap(NULL, sizeof(bool), PROT_READ | PROT_WRITE,  MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if ( MAP_FAILED == action_number) {
        return 1;
    }


    clerks_number = mmap(NULL, sizeof(bool), PROT_READ | PROT_WRITE,  MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if ( MAP_FAILED == clerks_number) {
        return 1;
    }

    customers_numbers = mmap(NULL, sizeof(bool), PROT_READ | PROT_WRITE,  MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if ( MAP_FAILED == customers_numbers) {
        return 1;
    }

    // inicializacia zdielanych premennych
    *post_is_closed = false;
    *first_service_queue = 0;
    *second_service_queue = 0;
    *third_service_queue = 0;
    *action_number = 0;
    *clerks_number = 0;
    *customers_numbers = 0;

    return 0;
}

// Customer process function
void customer_process(int idZ, int TZ) {
    (*customers_numbers)++;

    sem_wait(sem_mutex);
    // Print the initial message
    fprintf(file, "%d: Z %d: started\n", ++(*action_number), idZ);
    sem_post(sem_mutex);

    // Wait for a random time between 0 and TZ
    upsleep_for_random_time(TZ);

    sem_wait(sem_mutex);
    // Check if the post office is closed
    if (*post_is_closed) {
        fprintf(file, "%d: Z %d: going home\n", ++(*action_number), idZ);
        sem_post(sem_mutex);
        exit(0);
    }

    int service = rand() % NUM_SERVICES;
    fprintf(file, "%d: Z %d: entering office for a service %d\n", ++(*action_number), idZ, service + 1);
    sem_post(sem_mutex);

    switch(service) {
    case 0:
        (*first_service_queue)++;
        sem_wait(sem_first_service);
        break;
    case 1:
        (*second_service_queue)++;
        sem_wait(sem_second_service);
        break;
    case 2:
        (*third_service_queue)++;
        sem_wait(sem_third_service);
        break;
    default:
        printf("SERVICE ERROR\n");
        exit(1);
    }
    
    sem_wait(sem_mutex);
    fprintf(file, "%d: Z %d: called by office worker\n", ++(*action_number), idZ);
    sem_post(sem_mutex);

    upsleep_for_random_time(10);

    sem_wait(sem_mutex);
    fprintf(file, "%d: Z %d: going home\n", ++(*action_number), idZ);
    sem_post(sem_mutex);

    exit(0);
}

// Clerk process function
void clerk_process(int idU, int TU) {
    (*clerks_number)++;

    sem_wait(sem_mutex);
    fprintf(file, "%d: U %d: started\n", ++(*action_number), idU);
    sem_post(sem_mutex);

    while (1) {
        int services_queues[3] = {0, 0, 0};
        int service = -1;

        if (*first_service_queue)
        {   
            ++service;
            services_queues[service] = 1;
            
        }
        if (*second_service_queue)
        {
            ++service;
            services_queues[service] = 2;

        }    
        if (*third_service_queue)
        {
            ++service;
            services_queues[service] = 3;
        }

        if ((service == -1) && (*post_is_closed)) {
            sem_wait(sem_mutex);
            fprintf(file, "%d: U %d: going home\n", ++(*action_number), idU);
            sem_post(sem_mutex);
            (*clerks_number)--;
            exit(0);
        } else if (service == -1) {
            sem_wait(sem_mutex);
            fprintf(file, "%d: U %d: taking break\n", ++(*action_number), idU);
            sem_post(sem_mutex);

            upsleep_for_random_time(TU);

            sem_wait(sem_mutex);
            fprintf(file, "%d: U %d: break finished\n", ++(*action_number), idU);
            sem_post(sem_mutex);
        } else {
            

        










        }
        ////////////////////////////////

        
        


        //sem_wait(sem_mutex);
        int service = -1;
        for (int i = 0; i < NUM_SERVICES; i++) {
            if ((*customer_services_queue[i]) > 0) {
                service = i;
                //sem_post(sem_mutex);
                break;
            }
        }
        sem_wait(sem_mutex);

        sem_post(sem_mutex);
        if ((service == -1) && ((*post_is_closed) > 0)) {
            sem_wait(sem_mutex);
            fprintf(file, "%d: U %d: going home\n", ++(*action_number), idU);
            (*clerks_amount)--;
            sem_post(sem_mutex);
            exit(0);
        } else if (service == -1) {
            sem_wait(sem_mutex);
            fprintf(file, "%d: U %d: taking break\n", ++(*action_number), idU);
            sem_post(sem_mutex);

            upsleep_for_random_time(TU);

            sem_wait(sem_mutex);
            fprintf(file, "%d: U %d: break finished\n", ++(*action_number), idU);
            sem_post(sem_mutex);
            //continue;
        } else {

        //sem_post(sem_mutex);
        
        sem_post(sem_customer_services[service]);
        sem_wait(sem_customer_waiting); //waiting for customer to come.


        sem_wait(sem_mutex);
        fprintf(file, "%d: U %d: serving customer at service %d\n", ++(*action_number), idU, service + 1);
        (*customer_services_queue[service])--;
        sem_post(sem_mutex);

        usleep(rand() % 11);


        sem_wait(sem_mutex);
        fprintf(file, "%d: U %d: finished serving customer\n", ++(*action_number), idU);
        sem_post(sem_mutex);
        }
    }
    //return;
}

////////////////////////////    OTHER FUNCTIONS ////////////////////////////
// int random_number(int min, int max)
// {
//     int number = rand() % (max-min+1);
//     return number+min;
// }
