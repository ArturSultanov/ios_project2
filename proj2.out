1: Z 1: started
2: Z 1: entering office for a service 2
3: Z 2: started
4: Z 2: entering office for a service 1
5: Z 3: started
6: Z 3: entering office for a service 2
7: Z 4: started
8: Z 4: entering office for a service 3
9: Z 5: started
10: Z 5: entering office for a service 3
11: Z 6: started
12: Z 7: started
13: Z 8: started
14: Z 8: entering office for a service 2
15: Z 9: started
16: Z 10: started
17: Z 11: started
18: Z 11: entering office for a service 1
19: Z 12: started
20: Z 13: started
21: Z 14: started
22: Z 14: entering office for a service 3
23: Z 15: started
24: Z 15: entering office for a service 1
25: Z 16: started
26: Z 16: entering office for a service 1
27: Z 17: started
28: Z 17: entering office for a service 1
29: Z 18: started
30: Z 18: entering office for a service 3
31: Z 19: started
32: Z 20: started
33: Z 20: entering office for a service 1
34: Z 21: started
35: Z 22: started
36: Z 23: started
37: Z 23: entering office for a service 1
38: Z 24: started
39: Z 25: started
40: Z 26: started
41: Z 26: entering office for a service 1
42: Z 27: started
43: Z 28: started
44: Z 28: entering office for a service 3
45: Z 29: started
46: Z 29: entering office for a service 3
47: Z 30: started
48: Z 30: entering office for a service 2
49: U 1: started
50: Z 2: called by office worker
51: U 1: serving customer at service 1
52: U 1: finished serving customer
53: Z 11: called by office worker
54: U 1: serving customer at service 1
55: U 1: finished serving customer
56: U 2: started
57: Z 15: called by office worker
58: Z 15: going home
59: U 1: serving customer at service 1
60: Z 16: called by office worker
61: U 3: started
62: U 1: finished serving customer
63: U 2: serving customer at service 1
64: Z 17: called by office worker
65: Z 20: called by office worker
66: U 4: started
67: U 2: finished serving customer
68: U 3: serving customer at service 1
69: U 1: serving customer at service 1
70: Z 23: called by office worker
71: U 3: finished serving customer
72: Z 26: called by office worker
73: U 1: finished serving customer
74: U 4: serving customer at service 1
75: U 2: serving customer at service 1
76: U 5: started
77: U 4: finished serving customer
78: U 2: finished serving customer
79: Z 1: called by office worker
80: Z 8: called by office worker
81: U 6: started
82: Z 3: called by office worker
83: U 3: serving customer at service 1
84: U 1: serving customer at service 1
85: U 5: serving customer at service 2
86: Z 30: called by office worker
87: U 3: finished serving customer
88: U 1: finished serving customer
89: U 5: finished serving customer
90: U 4: serving customer at service 2
91: U 7: started
92: U 4: finished serving customer
93: U 8: started
94: U 9: started
95: U 10: started
96: U 11: started
97: U 12: started
98: U 13: started
99: U 14: started
100: Z 3: going home
101: U 15: started
102: U 16: started
103: U 17: started
104: U 18: started
105: U 19: started
106: U 20: started
107: Z 20: going home
108: Z 8: going home
109: Z 30: going home
110: Z 7: entering office for a service 2
111: Z 12: entering office for a service 2
112: Z 7: called by office worker
113: Z 12: called by office worker
114: U 2: serving customer at service 2
115: U 6: serving customer at service 2
116: U 2: finished serving customer
117: U 6: finished serving customer
118: Z 11: going home
119: Z 19: entering office for a service 2
120: Z 19: called by office worker
121: U 3: serving customer at service 2
122: U 3: finished serving customer
123: Z 23: going home
124: Z 16: going home
125: Z 17: going home
126: Z 24: entering office for a service 3
127: Z 1: going home
128: Z 27: entering office for a service 3
129: Z 22: entering office for a service 1
130: Z 2: going home
131: Z 22: called by office worker
132: U 1: serving customer at service 2
133: U 1: finished serving customer
134: Z 26: going home
135: Z 12: going home
136: Z 6: entering office for a service 2
137: Z 6: called by office worker
138: U 5: serving customer at service 2
139: U 5: finished serving customer
140: Z 19: going home
141: Z 9: entering office for a service 1
142: Z 13: entering office for a service 3
143: Z 9: called by office worker
144: U 7: serving customer at service 2
145: U 7: finished serving customer
146: Z 4: called by office worker
147: U 4: serving customer at service 2
148: Z 10: entering office for a service 1
149: U 4: finished serving customer
150: Z 10: called by office worker
151: Z 10: going home
152: U 8: serving customer at service 2
153: U 8: finished serving customer
154: Z 21: entering office for a service 3
155: Z 25: entering office for a service 1
156: Z 25: called by office worker
157: U 9: serving customer at service 2
158: U 9: finished serving customer
159: Z 7: going home
160: Z 4: going home
161: Z 25: going home
162: Z 22: going home
163: Z 9: going home
164: Z 6: going home
165: closing
