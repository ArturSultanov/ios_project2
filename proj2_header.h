// Artur Sultanov
// xsulta01
// IOS project 2 2023

#ifndef PROJ2_HEADER_FILE
#define PROJ2_HEADER_FILE

// Headers

#include <stdio.h>      // standard input/output functions
#include <stdbool.h>    // the bool type
#include <stdlib.h>     // functions like exit, malloc, and free

#include <unistd.h>     // the fork, usleep 
#include <sys/types.h>  // data type 'pid_t'
#include <time.h>       // time-related functions
#include <fcntl.h>      // flags like O_CREAT, O_EXCL
#include <semaphore.h>  // POSIX semaphore functions
#include <sys/wait.h>   // wait and waitpid functions
#include <sys/mman.h>   // shared memory functions
#include <signal.h>     // signal handling functions

#define SEMAPHORE_MUTEX "/xsulta01_sem_mutex"
#define SEMAPHORE_SERVICEFRST "/xsulta01_servicefrst" 
#define SEMAPHORE_SERVICESCND "/xsulta01_servicescnd" 
#define SEMAPHORE_SERVICETHRD "/xsulta01_servicethrd" 
#define SEMAPHORE_CLERK "/xsulta01_clerk"

#define NUM_SERVICES 3  // Number of Post services: dopisy:1, balíky:2, peněžní služby:3.
#define upsleep_for_random_time(time_max) usleep((rand() % (time_max + 1)) * 1000) // Macros for wait random time in range from 0 to time_max(inclusive).

// Function prototypes

int semaphore_init(void);               // Semaphores initialization(opening) function.
void semaphore_dest(void);              // Semaphores destruction(closing, unlinking) function.
int shared_memory_init(void);           // Shared memory initialization(mapping) function.
int shared_memory_dest(void);           // Shared memory destruction(unmapping) function.
void customer_process(int idZ, int TZ); // Customer-process logic.
void clerk_process(int idU, int TU);    // Clerk-process logic.
void cleanup(void);                     // Semaphores and shared variables destruction, closing output *file, if it wasn't.
void kill_child_processes(void);        // Kill all Child-processes were created by Main-process.
 
#endif
