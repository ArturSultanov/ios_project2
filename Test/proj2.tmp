1: Z 1: started
2: Z 2: started
3: Z 3: started
4: U 1: started
5: U 2: started
6: U 2: taking break
7: Z 3: entering office for a service 3
8: Z 2: entering office for a service 3
9: closing
10: Z 1: going home
11: U 2: break finished
12: U 1: serving a service of type 3
13: Z 3: called by office worker
14: U 2: serving a service of type 3
15: Z 2: called by office worker
16: U 1: service finished
17: U 2: service finished
18: U 1: going home
19: U 2: going home
20: Z 2: going home
21: Z 3: going home
