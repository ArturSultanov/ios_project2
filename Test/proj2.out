1: Z 1: started
2: Z 2: started
3: Z 3: started
4: Z 4: started
5: Z 5: started
6: Z 6: started
7: Z 7: started
8: Z 8: started
9: Z 9: started
10: Z 10: started
11: Z 11: started
12: Z 12: started
13: Z 13: started
14: Z 14: started
15: Z 15: started
16: Z 16: started
17: Z 17: started
18: Z 18: started
19: Z 19: started
20: Z 20: started
21: Z 21: started
22: Z 22: started
23: Z 23: started
24: Z 24: started
25: Z 25: started
26: Z 26: started
27: Z 27: started
28: Z 28: started
29: Z 29: started
30: Z 30: started
31: Z 31: started
32: Z 32: started
33: Z 33: started
34: Z 34: started
35: Z 35: started
36: Z 36: started
37: Z 37: started
38: Z 38: started
39: Z 39: started
40: Z 40: started
41: Z 41: started
42: Z 42: started
43: Z 43: started
44: Z 44: started
45: Z 45: started
46: Z 46: started
47: Z 47: started
48: Z 48: started
49: Z 49: started
50: Z 50: started
51: Z 51: started
52: Z 52: started
53: U 1: started
54: U 2: started
55: U 3: started
56: U 4: started
57: U 5: started
58: U 6: started
59: U 7: started
60: U 8: started
61: U 9: started
62: U 10: started
63: U 11: started
64: U 12: started
65: U 13: started
66: U 14: started
67: U 15: started
68: U 16: started
69: U 17: started
70: U 18: started
71: U 19: started
72: U 20: started
73: U 21: started
74: U 22: started
75: U 23: started
76: U 24: started
77: U 25: started
78: U 26: started
79: U 27: started
80: U 28: started
81: U 29: started
82: U 30: started
83: U 31: started
84: U 32: started
85: U 33: started
86: U 34: started
87: U 35: started
88: U 36: started
89: U 37: started
90: U 38: started
91: U 39: started
92: U 40: started
93: U 41: started
94: U 42: started
95: U 43: started
96: U 44: started
97: U 45: started
98: U 46: started
99: U 47: started
100: U 48: started
101: U 49: started
102: U 50: started
103: U 51: started
104: U 52: started
105: U 53: started
106: U 54: started
107: U 55: started
108: U 56: started
109: U 57: started
110: U 58: started
111: U 59: started
112: U 60: started
113: U 61: started
114: U 62: started
115: U 63: started
116: U 64: started
117: U 65: started
118: U 66: started
119: U 67: started
120: U 68: started
121: U 69: started
122: U 70: started
123: U 71: started
124: U 72: started
125: U 73: started
126: U 74: started
127: U 75: started
128: U 76: started
129: U 77: started
130: U 78: started
131: U 79: started
132: U 80: started
133: U 81: started
134: U 82: started
135: U 83: started
136: U 84: started
137: U 85: started
138: U 86: started
139: U 86: taking break
140: Z 4: entering office for a service 2
141: Z 15: entering office for a service 3
142: Z 35: entering office for a service 3
143: Z 2: entering office for a service 2
144: Z 27: going home
145: closing
146: Z 45: going home
147: Z 23: going home
148: Z 38: going home
149: Z 51: going home
150: Z 32: going home
151: Z 14: going home
152: Z 10: going home
153: U 86: break finished
154: U 1: serving a service of type 3
155: Z 15: called by office worker
156: U 2: serving a service of type 2
157: Z 4: called by office worker
158: U 1: service finished
159: U 3: serving a service of type 3
160: Z 35: called by office worker
161: U 5: serving a service of type 2
162: Z 2: called by office worker
163: U 4: going home
164: U 2: service finished
165: Z 4: going home
166: U 6: going home
167: U 7: going home
168: U 8: going home
169: U 9: going home
170: U 3: service finished
171: U 10: going home
172: U 11: going home
173: U 12: going home
174: U 13: going home
175: U 5: service finished
176: U 14: going home
177: U 15: going home
178: U 16: going home
179: U 17: going home
180: U 18: going home
181: U 19: going home
182: U 20: going home
183: U 21: going home
184: U 22: going home
185: U 23: going home
186: U 24: going home
187: U 25: going home
188: U 26: going home
189: U 27: going home
190: U 28: going home
191: U 29: going home
192: U 30: going home
193: U 31: going home
194: U 32: going home
195: U 33: going home
196: U 34: going home
197: U 35: going home
198: U 36: going home
199: U 37: going home
200: U 38: going home
201: U 39: going home
202: U 41: going home
203: U 42: going home
204: U 43: going home
205: U 44: going home
206: U 45: going home
207: U 40: going home
208: U 46: going home
209: U 47: going home
210: U 48: going home
211: U 49: going home
212: U 50: going home
213: U 51: going home
214: U 52: going home
215: U 53: going home
216: U 54: going home
217: U 55: going home
218: U 56: going home
219: U 57: going home
220: U 58: going home
221: U 59: going home
222: U 60: going home
223: U 61: going home
224: U 62: going home
225: U 63: going home
226: U 64: going home
227: U 65: going home
228: U 66: going home
229: U 67: going home
230: U 68: going home
231: U 69: going home
232: U 70: going home
233: U 71: going home
234: U 72: going home
235: U 73: going home
236: U 74: going home
237: U 75: going home
238: U 76: going home
239: U 77: going home
240: U 78: going home
241: U 79: going home
242: U 80: going home
243: U 81: going home
244: U 82: going home
245: U 83: going home
246: U 84: going home
247: U 85: going home
248: U 86: going home
249: U 1: going home
250: U 2: going home
251: U 3: going home
252: U 5: going home
253: Z 7: going home
254: Z 2: going home
255: Z 3: going home
256: Z 15: going home
257: Z 30: going home
258: Z 35: going home
259: Z 46: going home
260: Z 37: going home
261: Z 24: going home
262: Z 49: going home
263: Z 13: going home
264: Z 40: going home
265: Z 34: going home
266: Z 41: going home
267: Z 42: going home
268: Z 43: going home
269: Z 5: going home
270: Z 44: going home
271: Z 12: going home
272: Z 28: going home
273: Z 36: going home
274: Z 47: going home
275: Z 25: going home
276: Z 31: going home
277: Z 52: going home
278: Z 9: going home
279: Z 17: going home
280: Z 8: going home
281: Z 48: going home
282: Z 18: going home
283: Z 26: going home
284: Z 19: going home
285: Z 50: going home
286: Z 20: going home
287: Z 21: going home
288: Z 33: going home
289: Z 1: going home
290: Z 16: going home
291: Z 6: going home
292: Z 11: going home
293: Z 29: going home
294: Z 22: going home
295: Z 39: going home
